### PR TITLE
FEAT: Add Span Kind support for ES/OS

### PR DESCRIPTION
**Pull request by Manik2708**
## Which problem is this PR solving?
Fixes: #1923 

## Description of the changes
- While querying `GetOperations`, operations can now be fetched with kind also. When kind kept empty, spans of all kinds are returned

## How was this change tested?
- Unit and E2E tests

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `npm run lint` and `npm run test`

### CODE REVIEW

File: plugin/storage/es/spanstore/dbmodel/model.go [hunk 1]
```diff
@@ -88,3 +88,9 @@ type Service struct {
 	ServiceName   string `json:"serviceName"`
 	OperationName string `json:"operationName"`
 }
+
+// ServiceWithKind is the JSON struct service:kind:operation documents in ElasticSearch
+type ServiceWithKind struct {
+	Service
+	Kind string `json:"spanKind"`
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/dbmodel/model.go:95]: not sure I follow. Why wouldn't we just add Kind field to the Service struct above?

File: plugin/storage/es/spanstore/writer.go [hunk 2]
```diff
@@ -124,8 +124,9 @@ func getSpanAndServiceIndexFn(p SpanWriterParams) spanAndServiceIndexFn {
 func (s *SpanWriter) WriteSpan(_ context.Context, span *model.Span) error {
 	spanIndexName, serviceIndexName := s.spanServiceIndex(span.StartTime)
 	jsonSpan := s.spanConverter.FromDomainEmbedProcess(span)
+	kind, _ := span.GetSpanKind()
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/writer.go:127]: Why not add Kind field to the dbmodel.Span and not pass it around separately?

File: plugin/storage/es/spanstore/service_operation.go [hunk 3]
```diff
@@ -53,17 +57,28 @@ func NewServiceOperationStorage(
...
 	service := dbmodel.Service{
 		ServiceName:   jsonSpan.Process.ServiceName,
 		OperationName: jsonSpan.OperationName,
 	}
-
-	cacheKey := hashCode(service)
-	if !keyInCache(cacheKey, s.serviceCache) {
-		s.client().Index().Index(indexName).Type(serviceType).Id(cacheKey).BodyJson(service).Add()
-		writeCache(cacheKey, s.serviceCache)
+	if kind != model.SpanKindUnspecified {
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:66]: why do we need to bifurcate? If for some reason span kind is unavailable, we can just treat it as blank string and not change the code to deal with it separately.

**Review by yurishkuro — COMMENTED** [review]
overall lgtm, but I have question on whether the whole thing could be even simpler, by treating Kind as always present (but maybe blank).

**Comment by Manik2708** [comment]
> overall lgtm, but I have question on whether the whole thing could be even simpler, by treating Kind as always present (but maybe blank).

Then we can't get old data! In old data `Kind` was not present. In the actual issue, when it was asked that What will happen to old data, then it was answered that it should be accessible when kind is not present! Therefore I introduced a whole new struct with kind and saved the data without kind also.

File: plugin/storage/es/spanstore/service_operation.go (hunk 3 above)
Comment (Manik2708) [reply plugin/storage/es/spanstore/service_operation.go:66]: Please the comment https://github.com/jaegertracing/jaeger/pull/6399#issuecomment-2562068739. If the reason seems valid then I don't think we should introduce empty kinds because that will become messy along with data of `Without Kind` (old data) and then in reader it might become more complex while fetching.

File: plugin/storage/es/spanstore/writer.go (hunk 2 above)
Comment (Manik2708) [reply plugin/storage/es/spanstore/writer.go:127]: It will be present as a tag in json model also! I tried fetching kind from original span because of the bool `AllTagsAsFields`. As it will already be present in the json model, should we resave it as a seperate kind also?

**Comment by Manik2708** [comment]
> > overall lgtm, but I have question on whether the whole thing could be even simpler, by treating Kind as always present (but maybe blank).
> 
> Then we can't get old data! In old data `Kind` was not present. In the actual issue, when it was asked that What will happen to old data, it was answered that it should be accessible when kind is not present! Therefore I introduced a whole new struct with kind and saved the data without kind also.

File: plugin/storage/es/spanstore/service_operation.go [hunk 4]
```diff
@@ -127,9 +171,75 @@ func getOperationsAggregation(maxDocCount int) elastic.Query {
...
+		result[i] = spanstore.Operation{
+			Name:     str,
+			SpanKind: kind,
+		}
+	}
+	return result, nil
+}
+
+func bucketOfOperationsToOperationsArray(searchResult *elastic.AggregationBucketFilters) ([]spanstore.Operation, error) {
+	var result []spanstore.Operation
```
Comment (Manik2708) [inline plugin/storage/es/spanstore/service_operation.go:190]: This is concerning because we are unaware of the total length of result so not sure whether we should initialize this way or not. The best possible solution which could think of is as number of kinds are limited we can iterate through each kind fetch the total doc count of each kind, then finally create the array of total length.

File: plugin/storage/es/spanstore/dbmodel/model.go (hunk 1 above)
Comment (Manik2708) [reply plugin/storage/es/spanstore/dbmodel/model.go:95]: Please see the comment. As then in query for fetching spans of all kinds, we have apply another filter with kind as empty.

**Comment by Manik2708** [comment]
@yurishkuro Comitted as per your suggestions. Please review this

**Comment by yurishkuro** [comment]
please update the branch, I am not sure if CI is failing because of that or because of your changes. If it's the latter, how are you testing this change? Did you run e2e tests locally?

**Comment by Manik2708** [comment]
> please update the branch, I am not sure if CI is failing because of that or because of your changes. If it's the latter, how are you testing this change? Did you run e2e tests locally?

I tried approaching with empty kind but failed as query in `GetOperations` is becoming more complex.

**Comment by Manik2708** [comment]
@yurishkuro Saving empty strings in ES is not optimal. Please see: https://github.com/elastic/elasticsearch/issues/7515. I am thinking of saving it as `null` when kind is `Unspecified`

**Comment by yurishkuro** [comment]
I don't have a strong opinion, but the issue you linked talks about _searching_ for empty strings, which I don't think is the case for our scenario, we just need to write it.

**Comment by Manik2708** [comment]
> I don't have a strong opinion, but the issue you linked talks about _searching_ for empty strings, which I don't think is the case for our scenario, we just need to write it.

But while reading we have to search for those operations also which have empty kinds or no kind! I am facing problems when fetching spans of all kinds, the search query behaves weirdly if I introduce filter of `""`. There are no problems in writing the service but in fetching them

**Comment by yurishkuro** [comment]
Searching for "all kinds" to me means not specifying any filter for the "kind" field. How would it even work if you say `search for kind=""` when the kind is actually not empty?

**Comment by Manik2708** [comment]
> Searching for "all kinds" to me means not specifying any filter for the "kind" field. How would it even work if you say `search for kind=""` when the kind is actually not empty?

I think there is some sort of misunderstanding of my approach. Let me first explain problems:
1. We want to get two fields (operatioName and kind) from our search query to ES, to achieve this, we have 3 ways:
a) Use `FetchSource(true): This will fetch all the fields which is not optimal neither is avilable in our abstraction.
b) Use `FetchSourceWithContext`: Not available in our abstraction.
c) Composite Filters: Not possible because it will be applied to a field and kind can be absent in old data.
So I employed my approach, used the query only on service name and applied filters on kind (kinds are limited so it is possible) and used the filter name to get the kind from names buckets. But when this filter is applied to empty kind, it is not giving results. Your thoughts are absolutely correct when you said: When no kind is in query we have to just employ query on service name but we have to fetch two fields which is creating problem in just querying on service!

**Comment by yurishkuro** [comment]
"Not available in our abstraction" is an odd argument since we own the abstraction and can change it at will. Re fetchsource, what is the source here - is it the whole span? Or do we write separate entries just for service/operation?

**Comment by Manik2708** [comment]
> "Not available in our abstraction" is an odd argument since we own the abstraction and can change it at will. Re fetchsource, what is the source here - is it the whole span? Or do we write separate entries just for service/operation?

Service, Operation and Kind (if present)

**Comment by Manik2708** [comment]
Will try to clean this PR as soon as possible

**Comment by Manik2708** [comment]
@yurishkuro I have committed with `FetchSourceContext`, please review!

File: plugin/storage/es/spanstore/service_operation.go [hunk 5]
```diff
@@ -96,29 +100,46 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
...
+func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service, kind string, maxDocCount int) ([]spanstore.Operation, error) {
+	var searchService es.SearchService
+	if kind != "" {
+		serviceFilter := getOperationsAggregation(maxDocCount)
+		serviceNameQuery := elastic.NewTermQuery(serviceName, service)
+		serviceKindQuery := elastic.NewTermQuery(spanKind, kind)
+		serviceQuery := elastic.NewBoolQuery().Must(serviceNameQuery, serviceKindQuery)
+		searchService = s.client().Search(indices...).
+			Size(0).
+			Query(serviceQuery).
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:112]: the above decomposition is hard to follow, please inline it here with good formatting, e.g.
```suggestion
			Query(elastic.NewBoolQuery().Must(
				elastic.NewTermQuery(serviceName, service),
				elastic.NewTermQuery(spanKind, kind)
			)).
```

File: plugin/storage/es/spanstore/service_operation.go [hunk 6]
```diff
@@ -96,29 +100,46 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
...
-func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service string, maxDocCount int) ([]string, error) {
+func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service, kind string, maxDocCount int) ([]spanstore.Operation, error) {
+	var searchService es.SearchService
+	if kind != "" {
+		serviceFilter := getOperationsAggregation(maxDocCount)
+		serviceNameQuery := elastic.NewTermQuery(serviceName, service)
+		serviceKindQuery := elastic.NewTermQuery(spanKind, kind)
+		serviceQuery := elastic.NewBoolQuery().Must(serviceNameQuery, serviceKindQuery)
+		searchService = s.client().Search(indices...).
+			Size(0).
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:111]: what does Size(0) do? can you add a brief line comment?

File: plugin/storage/es/spanstore/service_operation.go [hunk 7]
```diff
@@ -96,29 +100,46 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
...
+			Query(serviceQuery).
+			IgnoreUnavailable(true).
+			Aggregation(operationsAggregation, serviceFilter)
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:114]: similarly, please inline arguments, it's very difficult to follow this query if I need to constantly jump elsewhere.


please rename `operationsAggregation` -> `operationsAggName`

File: plugin/storage/es/spanstore/service_operation.go [hunk 8]
```diff
@@ -96,29 +100,46 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
 		Size(maxDocCount) // ES deprecated size omission for aggregating all. https://github.com/elastic/elasticsearch/issues/18838
 }
 
-func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service string, maxDocCount int) ([]string, error) {
+func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service, kind string, maxDocCount int) ([]spanstore.Operation, error) {
+	var searchService es.SearchService
+	if kind != "" {
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:105]: please move into another function `getOperationsWithKind`

File: plugin/storage/es/spanstore/service_operation.go [hunk 9]
```diff
@@ -96,29 +100,46 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
...
-	serviceFilter := getOperationsAggregation(maxDocCount)
-
-	searchService := s.client().Search(indices...).
-		Size(0).
+	searchService = s.client().Search(indices...).
 		Query(serviceQuery).
 		IgnoreUnavailable(true).
-		Aggregation(operationsAggregation, serviceFilter)
-
+		FetchSourceContext(elastic.NewFetchSourceContext(true).Include(spanKind, operationNameField)).
```
Comment (yurishkuro) [inline plugin/storage/es/spanstore/service_operation.go:133]: I don't really understand this, can you please point to documentation of how this impacts the query behavior?

**Review by yurishkuro — COMMENTED** [review]
sorry for delay

File: plugin/storage/es/spanstore/service_operation.go (hunk 9 above)
Comment (Manik2708) [reply plugin/storage/es/spanstore/service_operation.go:133]: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-fields.html#field-retrieval-methods. This part of docs suggests to use fields over source, but I used `FetchSourceContext` because:
1. Weirdly olivere doesn't have `Fields` in their `SearchService` (maybe it's depreceated that's why) but rather it is in `ExplainService`.
2. Technically we need the whole source as Service mapping has only three fields: Service, Kind, Operation Name.

**Comment by Manik2708** [comment]
@yurishkuro Please can you review this!

**Comment by Manik2708** [comment]
Is this good to go? @yurishkuro

File: internal/storage/v1/elasticsearch/spanstore/service_operation.go [hunk 10]
```diff
@@ -96,29 +99,44 @@ func getServicesAggregation(maxDocCount int) elastic.Query {
 		Size(maxDocCount) // ES deprecated size omission for aggregating all. https://github.com/elastic/elasticsearch/issues/18838
 }
 
-func (s *ServiceOperationStorage) getOperations(ctx context.Context, indices []string, service string, maxDocCount int) ([]string, error) {
+func (s *ServiceOperationStorage) getOperationsWithKind(ctx context.Context, indices []string, service, kind string, maxDocCount int) ([]spanstore.Operation, error) {
+	var searchService es.SearchService
+	if kind != "" {
```
Comment (mahadzaryab1) [inline internal/storage/v1/elasticsearch/spanstore/service_operation.go:104]: this function is a bit difficult to read - can we move all the logic in this branch to `getOperationsWithKind` and leave this function to be called `getOperations`?

Comment (Manik2708) [reply internal/storage/v1/elasticsearch/spanstore/service_operation.go:104]: But then `getOperations` will be left unused! Should I still proceed forward?

**Comment by Manik2708** [comment]
@mahadzaryab1 Have completed the code-coverage. A humble reminder to review this PR!

**Comment by yurishkuro** [comment]
@Manik2708 I hate getting good work go to waste, but upon further reflection I rather doubt that the desired functionality is worth all this complexity. The ask in the ticket is to filter down the list of endpoints of a service by span kind. Most services only produce a small number of span names, my guess is p99.99 is less than 100. So the simplest solution is to load all operations from storage and just filter them down before returning to the client. Since you're just refactoring the span reader for ES, you can add another PR which includes a simple `if` statement in the loop when forming the response from GetOperations. Let's just do that after we merge the refactoring PR. I would then close this PR - I think it was a good exercise for you to learn about ES internals.

**Comment by Manik2708** [comment]
> @Manik2708 I hate getting good work go to waste, but upon further reflection I rather doubt that the desired functionality is worth all this complexity. The ask in the ticket is to filter down the list of endpoints of a service by span kind. Most services only produce a small number of span names, my guess is p99.99 is less than 100. So the simplest solution is to load all operations from storage and just filter them down before returning to the client. Since you're just refactoring the span reader for ES, you can add another PR which includes a simple `if` statement in the loop when forming the response from GetOperations. Let's just do that after we merge the refactoring PR. I would then close this PR - I think it was a good exercise for you to learn about ES internals.

Ok, no problem! I think we can first refactor the whole reader, then I will look into this.

**Comment by yurishkuro** [comment]
Actually, I might be wrong - the loop filter approach will only work if the storage is able to include the span kind in the response in the first place. If it's not stored at all next to the operation, then we'd still need storage changes.

**Comment by Manik2708** [comment]
> Actually, I might be wrong - the loop filter approach will only work if the storage is able to include the span kind in the response in the first place. If it's not stored at all next to the operation, then we'd still need storage changes.

@yurishkuro Yes, that's what even I was thinking. Actually I have a plan which can reduce the complexity. Currently we have to use the source becuase we are trying to fetch two fields but I have observed two things:
1. There are limited span kinds (probably 6)
2. Searching is never done on operation name.

We have to eventually add one more field (I can't find any alternative) but we can use the approach of badger in ElasticSearch, that is add a byte at the end of operation name. So we can seggregate the old data without spankind field and send the operation name as it is and that having span kind, we can fetch span kind by the last byte. This way we can use the trivial queries without dealing with the source. Will be happy to get your views over this.
